### ISSUE TITLE
Should networks and SPA modules stay together?

### ISSUE THREAD

**Issue by tbekolay** [nengo/nengo#1144]
Github recently made is possible to [change the base branch of a pull request](https://github.com/blog/2224-change-the-base-branch-of-a-pull-request), so I've changed all of the open [SPA-labelled PRs](https://github.com/nengo/nengo/pulls?q=is%3Apr+label%3ASPA+is%3Aopen) to be based on the `new-spa` branch rather than `master`.

In doing so, however, I noticed a few cases in which the SPA PRs touch non-SPA code. Specifically:

1. #743 modifies `nengo/utils/numpy.py`
2. #810 modifies `nengo/networks/actionselection.py`
3. #991 modifies `nengo/networks/assoc_mem.py`
4. #1104 modifies `nengo/networks/circularconvolution.py`

I think the change in #743 is clearly something that should be split off into a separate PR, but the other three instances all touch the non-SPA versions of networks. So the question that I'd like people's opinions on is whether these networks should live in the new SPA module or whether they should remain in core Nengo?

**Comment by tbekolay** [comment]
My thoughts: there was some talk previously about splitting off all or most of the networks in Nengo to another repo, so it would make sense to me that the underlying networks that SPA makes would move along with the modules to be in the new SPA repo. If we add some infrastructure in Nengo or Nengo GUI to detect installed networks, then the new SPA repo would use that mechanism like any other repo. So it seems to me that there aren't any significant downsides to moving them with the rest of the SPA module.

**Comment by jgosmann** [comment]
I think we might have three sets of networks:

1. Core networks that should stay in core Nengo.
2. Extra networks that could be moved to an extra package (nengo_networks or nengo_extra or multiple packages).
3. SPA related networks.

Regarding 2. (extra networks) I wonder what would be the advantages or disadvantages of keeping those networks with core Nengo?

Regarding 3. (SPA networks)  it would probably make sense to move them to nengo_spa (even if they could be used without SPA modules, they will mainly used in that context).

Looking at our list of networks, I would them sort in the following way:

**Core**
CircularConvolution (not entirely sure, where to sort this; it is a pretty basic operation and used a lot, but mostly in the context of SPA, but not dependent on it)
EnsembleArray
Integrator
Oscillator
Product

**Extra**
InputGatedMemory

**SPA**
actionselection (BasalGanglia, Thalamus)
AssociativeMemory


One worry about too much fragmantation is that it makes it hard to discover existing networks. If they are all in one place, there's only one place to check in.

**Comment by tbekolay** [comment]
Dev meeting update: we'll go with @jgosmann's proposal! We'll try to keep what's in core minimal (maybe just `EnsembleArray` and `Product`?) and migrate the others. SPA networks will go with SPA module (and leave core with `nengo.spa` once it's been deprecated for long enough).

**Comment by xchoo** [comment]
Hmm. Just a note that there are 2 versions of the AssociativeMemory network. One specifically tweaked for SPA use (i.e. it uses vocabularies), and one that's a generic network (i.e. it uses a list of vectors).

**Comment by tbekolay** [comment]
> 2 versions of the AssociativeMemory network

Yep, that's why I made this discussion issue in the first place, to think about whether both or just one should go with the SPA stuff when it splits off. Consensus was both; means that you might have to install `nengo_spa` even if you use the non-vocab version of it (if that exists in the end), but that's preferable versus having to keep the two networks in sync and tested if they're in different repos.

**Comment by hunse** [comment]
I'm not sure what you guys discussed about `CircularConvolution`, but I would move it to `nengo_spa` as well. As @tbekolay said, there could well be non-vocab versions in there, but would one ever want to do circular convolution of things that aren't semantic pointers in the broad sense? i.e. would anyone use it as anything but the binding operation in a vector-symbolic architecture? I think such cases would be rare.

**Comment by tbekolay** [comment]
> I would move CircularConvolution to nengo_spa as well

Yep, that was the consensus. Thought convergence! :rainbow:

**Comment by xchoo** [comment]
Hmmm. If we move the non-spa stuff into `nengo_spa`, perhaps we should consider `nengo_spa` to be more like `nengo_vsa` (I can see potentially using that repo for non-spa related stuff as well)?
